cmd/rook/ceph/operator.go

```diff
@@ -91,14 +111,132 @@ func startOperator(cmd *cobra.Command, args []string) error {
+	if err != nil {
+		// If secret is not found. All good ! Proceed with rook without admission controllers
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, errors.Errorf("Error happened while retrieving the Secret: %+v", err)
+	}
+	return true, nil
+}
+
+func initWebhook(context *clusterd.Context) error {
+	// At this point volume should be mounted, so proceed with creating the service and validatingwebhookconfig
+	err := createWebhookService(context)
+	if err != nil {
+		return errors.Errorf("failed to create service: %+v", err)
+	}
+	err = createValidatingWebhookConfig(context)
+	if err != nil {
+		return errors.Errorf("failed to create validatingwebhookconfig: %+v", err)
+	}
+	// As http.ListenAndServe is a blocking call. We are starting our server in its own goroutine so that rook functionality is not affected.
+	go webhook.StartServer(context)
+
+	return nil
+}
+
+func createWebhookService(context *clusterd.Context) error {
+	serviceClient := context.Clientset.CoreV1().Services(namespace)
+
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      acServiceName,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Port: servicePort,
+					TargetPort: intstr.IntOrString{
+						IntVal: serverPort,
+					},
+				},
+			},
+			Selector: map[string]string{
+				"app": "rook-ceph-operator",
+			},
+		},
+	}
+	_, err := serviceClient.Create(&webhookService)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func createValidatingWebhookConfig(context *clusterd.Context) error {
+	configMap, err := context.Clientset.CoreV1().ConfigMaps(kubeNamespace).Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Errorf("error with fetching configmap: %+v", err)
+	}
+
+	webhookClient := context.Clientset.AdmissionregistrationV1().ValidatingWebhookConfigurations()
+
+	sideEffects := admissionregistrationv1.SideEffectClassNone
+	validatingWebhookConfig := admissionregistrationv1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      webhookConfigName,
+			Namespace: namespace,
+		},
+		Webhooks: []admissionregistrationv1.ValidatingWebhook{
+			{
+				Name: acServiceName + "." + namespace + ".svc",
+				ClientConfig: admissionregistrationv1.WebhookClientConfig{
+					CABundle: []byte(configMap.Data["client-ca-file"]),
+					Service: &admissionregistrationv1.ServiceReference{
+						Namespace: namespace,
+						Name:      acServiceName,
+						Path:      &validatePath,
+					},
+				},
+				Rules: []admissionregistrationv1.RuleWithOperations{
+					{
+						Operations: []admissionregistrationv1.OperationType{
+							"CREATE",
+							"UPDATE",
+							"DELETE",
+						},
+						Rule: admissionregistrationv1.Rule{
+							APIGroups:   []string{"ceph.rook.io"},
+							APIVersions: []string{"v1"},
+							Resources:   []string{"cephclusters"},
```

Comment:
How about if we add all the rook CRDs to the list?

Reply:
I have added a few. Please let me know if there are more.